### PR TITLE
task/S20-442-proxy-backend

### PR DESCRIPTION
**Pull request by taaraora**
Proxy is implemented.
Default proxy port range is 60200 - 60250

### CODE REVIEW

File: pkg/kube/handler.go [hunk 1]
```diff
@@ -979,6 +986,7 @@ func (h *Handler) getNodesMetrics(w http.ResponseWriter, r *http.Request) {
 }
 
 func (h *Handler) getServices(w http.ResponseWriter, r *http.Request) {
+	fmt.Println("OLOLO")
```
Comment (egopher) [inline pkg/kube/handler.go]: please remove it.

File: pkg/kube/proxy.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+func (p *APIProxy) SetServices(kubeID string, services []*ServiceInfo) error {
+	if p.svc == nil || p.k8sClusterIDToProxies == nil {
+		return errors.New("component was not initialized properly")
+	}
+
+	p.servicesMux.Lock()
+
+	_, exists := p.k8sClusterIDToProxies[kubeID]
+	if !exists {
+		p.k8sClusterIDToProxies[kubeID] = make(map[string]*ServiceProxy)
```
Comment (egopher) [inline pkg/kube/proxy.go]: This will cause race conditions in case of concurrent requests. Please either add a mutex there or use sync.Map.

File: cmd/controlplane/main.go [hunk 3]
```diff
@@ -8,20 +8,24 @@ import (
...
-	uiDir         = flag.String("ui-dir", "./cmd/ui/assets/dist", "directory for supergiant ui to be served")
+	addr                 = flag.String("address", "0.0.0.0", "network interface to attach server to")
+	port                 = flag.Int("port", 8080, "tcp port to listen for incoming requests")
+	etcdURL              = flag.String("etcd-url", "localhost:2379", "etcd url with port")
+	templatesDir         = flag.String("templates", "/etc/supergiant/templates/", "supergiant will load script templates from the specified directory on start")
+	logLevel             = flag.String("log-level", "INFO", "logging level, e.g. info, warning, debug, error, fatal")
+	logFormat            = flag.String("log-format", "txt", "logging format [txt json]")
+	spawnInterval        = flag.Int("spawnInterval", 5, "interval between API calls to cloud provider for creating instance")
+	uiDir                = flag.String("ui-dir", "./cmd/ui/assets/dist", "directory for supergiant ui to be served")
+	ProxiesPortRangeFrom = flag.Int("proxies-port-from", 60200, "first tcp port in a range of binding reverse proxies for service apps")
```
Comment (saheienko) [inline cmd/controlplane/main.go]: Please, use a single flag for setting a port range.

example: https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/options/options.go#L185

File: pkg/kube/handler.go [hunk 4]
```diff
@@ -1003,12 +1009,12 @@ func (h *Handler) getServices(w http.ResponseWriter, r *http.Request) {
...
 		kubeID      string
 		masterNode  *node.Node
-		k8sServices = &K8SServices{}
+		k8sServices = &k8SServices{}
 	)
 	vars := mux.Vars(r)
 	kubeID = vars["kubeID"]
 
-	k, err := h.svc.Get(r.Context(), kubeID)
+	k8sCluster, err := h.svc.Get(r.Context(), kubeID)
```
Comment (saheienko) [inline pkg/kube/handler.go]: Please, use short names for local variables.
https://github.com/golang/go/wiki/CodeReviewComments#variable-names

File: pkg/kube/handler.go [hunk 5]
```diff
@@ -1057,122 +1063,85 @@ func (h *Handler) getServices(w http.ResponseWriter, r *http.Request) {
...
+				ID:        service.Metadata.UID,
+				Name:      service.Metadata.Name,
+				Type:      service.Spec.Type,
+				Namespace: service.Metadata.Namespace,
+			}
+			serviceInfos = append(serviceInfos, serviceInfo)
+
+			targetServices = append(targetServices, &proxy.Target{
+				kubeID + service.Metadata.UID,
+				fmt.Sprintf("https://%s%s:%d/proxy",
```
Comment (saheienko) [inline pkg/kube/handler.go]: Please, set fields explicitly.
```
struct{f1: v1, f2: v2} instead of struct{v1, v2} 
```

File: pkg/kube/handler.go [hunk 6]
```diff
@@ -1057,122 +1063,85 @@ func (h *Handler) getServices(w http.ResponseWriter, r *http.Request) {
...
+						ID:        service.Metadata.UID,
+						Name:      service.Metadata.Name,
+						Type:      service.Spec.Type,
+						Namespace: service.Metadata.Namespace,
 					}
-					services = append(services, service)
+					serviceInfos = append(serviceInfos, serviceInfo)
+
+					targetServices = append(targetServices, &proxy.Target{
+						kubeID + service.Metadata.UID,
```
Comment (saheienko) [inline pkg/kube/handler.go]: The unique identifier for services is namespace and name.

`UID` is the unique in time and space value for this object(if service is updated it will get another id).
https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#uids

File: pkg/kube/handler.go [hunk 7]
```diff
@@ -1057,122 +1063,85 @@ func (h *Handler) getServices(w http.ResponseWriter, r *http.Request) {
...
-	vars := mux.Vars(r)
-	kubeID := vars["kubeID"]
-
-	k, err := h.svc.Get(r.Context(), kubeID)
-	if err != nil {
-		if sgerrors.IsNotFound(err) {
-			message.SendNotFound(w, kubeID, err)
-			return
+func labelIsTrue(labelName string, labels map[string]string) bool {
+	for labelKey, labelValue := range labels {
```
Comment (saheienko) [inline pkg/kube/handler.go]: No need to iterate over a map:
```
v, ok := labels[name]
```
Also, could it more generic, like `contains(name, value, labels)`?

File: pkg/proxy/proxy.go [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+				logger.Errorf("error while serving address: %v", listener.Addr())
+			}
+		}
+	}()
+
+	logger.Infof("proxy server started on: %s, for targetURL: %+v", addr, targetURL)
+
+	return &ServiceReverseProxy{
+		srv:         httpServer,
+		servingBase: addr,
```
Comment (saheienko) [inline pkg/proxy/proxy.go]: Does `serviceBase` is a good name for server address?

File: pkg/proxy/proxy.go [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+	}, nil
+}
+
+func (p *ReverseProxyContainer) RegisterProxies(targets []*Target) error {
+	if p.Proxies == nil {
+		return errors.New("component was not initialized properly")
```
Comment (saheienko) [inline pkg/proxy/proxy.go]: Please, add component name to the error message.

File: pkg/proxy/proxy.go [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+		if strings.HasPrefix(inputURL, selfLink) {
+			inputURL = strings.TrimPrefix(inputURL, selfLink)
+		}
+
+		if strings.Index(inputURL, ":") == 0 {
+			parts := strings.Split(inputURL, "/")
+			inputURL = strings.TrimPrefix(inputURL, parts[0]+"/proxy")
+		}
+
+		logger.Infof("req.URL: %+v, inputURL %s, inputFullURL: %s",
```
Comment (saheienko) [inline pkg/proxy/proxy.go]: It's a kind of a spamming message, use `debug` logging level.

File: pkg/proxy/proxy.go [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+		// Update the headers to allow for SSL redirection
+		req.Header.Set("X-Forwarded-Host", req.Header.Get("Host"))
+		req.URL.Path = inputURL
+
+		// Note that ServeHttp is non blocking and uses a go routine under the hood
+		reverseProxy.ServeHTTP(res, req)
+	}
+}
+
+func (p *ReverseProxyContainer) Shutdown(ctx context.Context) {
```
Comment (saheienko) [inline pkg/proxy/proxy.go]: Is is possible to shut down some particular proxy?

File: templates/prometheus.sh.tpl [hunk 12]
```diff
@@ -2,8 +2,6 @@
 # it explicitly
 sudo /opt/bin/helm install stable/prometheus-operator \
     --name=prometheus-operator \
-    --set prometheus.prometheusSpec.externalUrl=/api/v1/namespaces/default/services/prometheus-operated:9090/proxy/ \
```
Comment (saheienko) [inline templates/prometheus.sh.tpl]: How does it related to service proxy?
Please, make a separate PR if it doesn't related.

File: pkg/kube/handler.go (hunk 4 above)
Comment (taaraora) [reply pkg/kube/handler.go]: Sorry, that is not my scenario.

File: pkg/kube/handler.go (hunk 5 above)
Comment (taaraora) [reply pkg/kube/handler.go]: done

File: pkg/kube/handler.go (hunk 6 above)
Comment (taaraora) [reply pkg/kube/handler.go]: I didn't get your point.

File: pkg/kube/handler.go (hunk 7 above)
Comment (taaraora) [reply pkg/kube/handler.go]: Oh, nice catch, thank you!

File: pkg/kube/handler.go (hunk 6 above)
Comment (saheienko) [reply pkg/kube/handler.go]: Don't use `service.Metadata.UID` as unique service identifier.

File: cmd/controlplane/main.go [hunk 13]
```diff
@@ -8,20 +8,24 @@ import (
 	"syscall"
 	"time"
 
+	"github.com/supergiant/control/pkg/proxy"
```
Comment (egopher) [inline cmd/controlplane/main.go]: gofmt

File: pkg/controlplane/server.go [hunk 14]
```diff
@@ -4,9 +4,13 @@ import (
 	"context"
 	"fmt"
 	"net/http"
+	"net/url"
 	"os"
+	"strings"
 	"time"
 
+	"github.com/supergiant/control/pkg/proxy"
```
Comment (egopher) [inline pkg/controlplane/server.go]: remove whitespace and gofmt

File: templates/prometheus.sh.tpl (hunk 12 above)
Comment (egopher) [reply templates/prometheus.sh.tpl]: It breaks service proxy.
